Code/GraphMol/Descriptors/GETAWAY.cpp

```diff
@@ -0,0 +1,804 @@
+std::vector<double>  clusterArray(std::vector<double> data) {
+    std::vector<double> Store;
+
+    // sort the input data
+    std::sort(data.begin(), data.end());
+
+    // find the difference between each number and its predecessor
+    std::vector<double> diffs;
+
+    std::adjacent_difference(data.begin(), data.end(), std::back_inserter(diffs));
+
+    // convert differences to percentage changes
+    std::transform(diffs.begin(), diffs.end(), data.begin(), diffs.begin(),
+        std::divides<double>());
+
+    int j=0;
+    int count=0;
+    for (unsigned int i = 0; i < data.size(); i++) {
+        count++;
+        // if a difference exceeds 1%, start a new group:
+        if (diffs[i] > 0.001)  {// diff=0.01 <=> 1%
+            Store.push_back(count);
+            count=0;
+            j++;
+        }
+    }
+
+    return Store;
+}
+
+
+double* GetGeodesicMatrix(double* dist, int lag,int numAtoms){
+    int sizeArray=numAtoms*numAtoms;
+    double *Geodesic = new double[sizeArray];
+    for (int i=0; i<sizeArray;i++) {
+      if (dist[i]==lag) Geodesic[i]=1;
+      else  Geodesic[i]=0;
+    }
+
+    return Geodesic;
+}
+
+
+JacobiSVD<MatrixXd> getSVD(MatrixXd A) {
+    JacobiSVD<MatrixXd> mysvd(A,  ComputeThinU | ComputeThinV);
+    return mysvd;
+}
+
+
+MatrixXd GetPinv(MatrixXd A){
+    JacobiSVD<MatrixXd> svd = getSVD(A);
+    double  pinvtoler=1.e-2; // choose your tolerance wisely!
+    VectorXd vs=svd.singularValues();
+    VectorXd vsinv=svd.singularValues();
+
+    for (unsigned int i=0; i<A.cols(); ++i) {
+        if ( vs(i) > pinvtoler )
+           vsinv(i)=1.0/vs(i);
+       else vsinv(i)=0.0;
+    }
+
+    MatrixXd S =  vsinv.asDiagonal();
+    MatrixXd Ap = svd.matrixV() * S * svd.matrixU().transpose();
+    return Ap;
+}
+
+
+MatrixXd GetCenterMatrix(MatrixXd Mat){
+    VectorXd v = Mat.colwise().mean();
+    MatrixXd X=Mat.rowwise() - v.transpose();
+    return X;
+}
+
+
+MatrixXd GetHmatrix(MatrixXd X){
+    MatrixXd  Weigthed = X.transpose()*X;
+    return X * GetPinv(Weigthed) * X.transpose();
+}
+
+
+MatrixXd GetRmatrix(MatrixXd H, MatrixXd DM, int numAtoms){
+
+    MatrixXd R = MatrixXd::Zero(numAtoms,numAtoms);
+    for (int i=0;i<numAtoms-1;i++) {
+      for (int j=i+1;j<numAtoms;j++){
+          R(i,j)=sqrt(H(i,i)*H(j,j)) / DM(i,j) ;
+          R(j,i)=R(i,j);
+      }
+    }
+
+    return R;
+}
+
+std::vector<int> GetHeavyList(const ROMol& mol){
+
+
+   int numAtoms= mol.getNumAtoms();
+
+    std::vector<int> HeavyList;
+   for (int i = 0; i < numAtoms; ++i) {
+
+          const RDKit::Atom * atom= mol.getAtomWithIdx(i);
+
+        if (atom->getAtomicNum() > 1) {
+            HeavyList.push_back(1);
+         }
+        else  HeavyList.push_back(0);
+    }
+  return HeavyList;
+}
+
+
+double* AppendDouble(double *w, double* Append, int length, int pos){
+    for (int i=pos;i<pos+length;i++) {
+          w[i]=Append[i-pos];
+        }
```

Comment:
It's essentially `concatenateVectors` and you have `vector1` and `vector2`. But it's not really appending, isn't it? This function copies the content of `vector2` into a `vector1`. vector1 must already have the allocated space. So maybe rather something like `copyIntoVector`.